docs/proposals/20210524-managed-external-etcd.md

```diff
@@ -0,0 +1,570 @@
+- Once the EtcdadmCluster is Initialized, the etcd cluster controller will scale up the cluster to add as many members as required to match EtcdadmCluster.Spec.Replicas field
+- At the end of every reconciliation loop, the etcd cluster controller will check if desired number of members are created by
+    * Getting Machines owned by EtcdadmCluster
+    * Running healthcheck for each Machine with the etcd `/health` endpoint at `https://client-url:2379/health`
+- Once all members are ready, EtcdadmCluster will set the endpoints on EtcdadmCluster.Status.Endpoints field, and set EtcdadmCluster.Status.CreationComplete to true
+- The CAPI cluster controller will have a new `reconcileEtcdCluster` function, that will check if EtcdadmCluster is ready, and when it is ready it will resume the control plane by deleting the "paused" annotation.
+- The control plane controller starts provisioning once it is no longer paused. It will first check if the Cluster spec contains the managedExternalEtcd ref. If the reference exists, it will get the endpoint from the referred etcdCluster object's status field. It will then start the provisioning of the control plane cluster using this etcd cluster.
+
+###### Upgrade or member replacement use case
+
+- The EtcdadmCluster.Spec.Replicas field will be immutable. This is to protect the etcd cluster from losing quorum by unnecessarily adding/removing healthy members.
+- An etcd member will need to be replaced by the etcd controller in two cases:
+  - Member is found unhealthy
+  - During an upgrade
+- While replacing an unhealthy etcd member, it is important to first remove the member and only then replace it by adding a new one. This ensures the cluster stays stable. [This Etcd doc](https://etcd.io/docs/v3.4/faq/#should-i-add-a-member-before-removing-an-unhealthy-member) explains why removing an unhealthy member is important in depth. Etcd's latest versions allow adding new members as learners so it doesn't affect the quorum size, but that feature is in beta. Also, adding a learner is a two-step process, where you first add a member as learner and then promote it, and etcdadm currently doesn't support it. So we can include this in a later version once the learner feature becomes GA and etcdadm incorporates the required changes.
+- We can reuse this same "scale down" function when removing a healthy member during an etcd upgrade. So even upgrade will follow the same pattern of removing the member first and then adding a new one with the latest spec.
+- Once the upgrade/machine replacement is completed, the etcd controller will again update the EtcdadmCluster.Status.Endpoints field. The control plane controller will rollout new Machines that use the updated etcd endpoints for the apiserver.
```

Comment:
Not sure I follow how do we preserve state during the upgrade operation.
Also, does this means that the control plane will work with stale list of endpoints while the etcd cluster is upgrading?